cmd/zed/auth/login.go

```diff
@@ -51,42 +56,55 @@ func (c *LoginCommand) Run(args []string) error {
 	switch method.Kind {
 	case api.AuthMethodAuth0:
 	case api.AuthMethodNone:
-		return fmt.Errorf("Zed lake service at %s does not support authentication", c.Lake)
+		return fmt.Errorf("Zed service at %s does not support authentication", c.Lake)
 	default:
-		return fmt.Errorf("Zed lake service at %s requires unknown authentication method %s", c.Lake, method.Kind)
+		return fmt.Errorf("Zed service at %s requires unknown authentication method %s", c.Lake, method.Kind)
 	}
 	fmt.Println("method", method.Auth0.ClientID)
 	fmt.Println("domain", method.Auth0.Domain)
 	fmt.Println("audience", method.Auth0.Audience)
-	dar, err := devauth.DeviceAuthorizationFlow(ctx, devauth.Config{
-		Audience: method.Auth0.Audience,
-		Domain:   method.Auth0.Domain,
-		ClientID: method.Auth0.ClientID,
-		Scope:    "openid profile email offline_access",
-		UserPrompt: func(res devauth.UserCodePrompt) error {
-			fmt.Println("Complete authentication at:", res.VerificationURL)
-			fmt.Println("User verification code:", res.UserCode)
-			if c.launchBrowser {
-				browser.OpenURL(res.VerificationURL)
-			}
-			return nil
-		},
-	})
+	auth0client, err := auth0.NewClient(*method.Auth0)
 	if err != nil {
 		return err
 	}
-	creds, err := c.LoadCredentials()
+	dcr, err := auth0client.GetDeviceCode(ctx, "openid email profile offline_access")
 	if err != nil {
-		return fmt.Errorf("failed to load credentials file: %w", err)
+		return err
 	}
-	creds.AddTokens(c.Lake, cli.ServiceTokens{
-		Access:  dar.AccessToken,
-		ID:      dar.IDToken,
-		Refresh: dar.RefreshToken,
-	})
-	if err := c.SaveCredentials(creds); err != nil {
+	fmt.Println("Complete authentication at", dcr.VerificationURIComplete)
+	fmt.Println("User verification code:", dcr.UserCode)
+	if c.launchBrowser {
+		browser.OpenURL(dcr.VerificationURIComplete)
+	}
+	tokens, err := c.pollForTokens(ctx, auth0client, dcr)
+	if err != nil {
+		return err
+	}
+	if err := c.AuthStore().SetLakeTokens(c.Lake, tokens); err != nil {
 		return fmt.Errorf("failed to save credentials file: %w", err)
 	}
 	fmt.Printf("Login successful, stored credentials for %s\n", c.Lake)
 	return nil
 }
+
+func (c *LoginCommand) pollForTokens(ctx context.Context, client *auth0.Client, dcr auth0.DeviceCodeResponse) (auth0.Tokens, error) {
```

Comment:
This doesn't make much sense as a `LoginCommand` method since the receiver is unused. But I think it does make sense as an `auth0.Client` method.

Reply:
Yeah, good idea.